python/kotekan/shared_memory_buffer.py

```diff
@@ -0,0 +1,539 @@
+        self._check_for_identical_timestamps(last_ts)
+
+        # return timestamps newer than the given one
+        ll = [i for i, v in enumerate(last_ts) if v > timestamp]
+        logger.debug(
+            "Found {} time slots that are newer than {} in {}: {}.".format(
+                len(ll), timestamp, last_ts, ll
+            )
+        )
+        return ll
+
+    def _return_data_copy_since(self, timestamp):
+        if not self._time_index_map:
+            return None
+
+        # sort time index map
+        last_ts, idxs = list(zip(*sorted(self._time_index_map.items())))
+
+        # get indexes of timestamps newer than the given on
+        idxs = [i for i, v in zip(idxs, last_ts) if v > timestamp]
+        logger.debug(
+            "Found {} time slots that are newer than {} in {}: {}.".format(
+                len(idxs), timestamp, last_ts, idxs
+            )
+        )
+
+        # select last n time slots
+        return VisRaw.from_nparray(
+            self._data[idxs, :],
+            self.size_frame,
+            len(idxs),
+            self.num_freq,
+            num_elements=7,
+            num_stack=28,
+            num_ev=0,
+        )
+
+    def _copy_from_shm(self, times, access_record):
+        # copy data updates within the last n time slots
+        for t in times:
+            for f_i in range(self.num_freq):
+                # check if this value should be copied: only if the access record changed and
+                # is not set to invalid
+                if access_record[t][f_i] != self.invalid_value and (
+                    self._last_access_record is None
+                    or access_record[t][f_i] > self._last_access_record[t][f_i]
+                ):
+                    logger.debug("Copying value at time={}, freq={}".format(t, f_i))
+                    # check if this time slot is in local copy of data
+                    try:
+                        t_i = self._time_index_map[access_record[t][f_i]]
+                    except KeyError:
+                        if len(self._time_index_map) == self.buffer_size:
+                            t_i = self._remove_oldest_time_slot()
+                        else:
+                            t_i = self._find_free_time_slot()
+                            if t_i is None:
+                                raise SharedMemoryError(
+                                    "Can't find a free slot in buffer (has keys {}).".format(
+                                        self._time_index_map.keys()
+                                    )
+                                )
+                        logger.debug(
+                            "Setting time index map [{}] to {}.".format(
+                                access_record[t][f_i], t_i
+                            )
+                        )
+                        self._time_index_map[access_record[t][f_i]] = t_i
+
+                    # copy the value
+                    # TODO: eliminate extra copy
```

Comment:
You're probably good as I'm pretty sure this doesn't actually copy.

Reply:
Yeah. This does not copy. The numpy docs are not super clear, but here's a quick experiment:
```
>>> a = np.arange(10)
>>> print(np.ndarray(3, np.int64, a, 8, order="C").flags)
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
```
Note that `OWNDATA` is False.

Reply:
Here I actually want a copy. If I get a view, it can change after I validated the access record. But doesn't
```
self._data[t_i, f_i] = tmp[:]
```
copy?

Reply:
Yup. That does copy. Sorry, if this wasn't clear, I was just inferring (perhaps mistakenly) that the `TODO` above was essentially about whether `tmp` creates and extra copy, and I'm saying that you're good, as it creates a view, and then the assignment below creates a copy.